packages/big-design/src/components/ButtonGroup/ButtonGroup.tsx

```diff
@@ -0,0 +1,153 @@
+import { MoreHorizIcon } from '@bigcommerce/big-design-icons';
+import React, { createRef, HTMLAttributes, memo, useCallback, useEffect, useMemo, useState } from 'react';
+
+import { useWindowResizeListener } from '../../hooks';
+import { MarginProps } from '../../mixins';
+import { ButtonProps } from '../Button';
+import { Dropdown } from '../Dropdown';
+import { Flex } from '../Flex';
+
+import { StyledButton, StyledFlexItem } from './styled';
+
+export interface ButtonGroupAction extends Omit<ButtonProps, 'children' | 'iconOnly' | 'iconRight' | 'iconLeft'> {
+  text: string;
+  icon?: React.ReactElement;
+}
+
+export interface ButtonGroupProps extends HTMLAttributes<HTMLDivElement>, MarginProps {
+  actions: ButtonGroupAction[];
+}
+
+const excludeIconProps = <T extends any>({
+  iconOnly,
+  iconRight,
+  iconLeft,
+  ...actionProps
+}: T): Pick<T, Exclude<keyof T, 'iconOnly' | 'iconRight' | 'iconLeft'>> => actionProps;
+
+export const ButtonGroup: React.FC<ButtonGroupProps> = memo(({ actions, ...wrapperProps }) => {
+  const parentRef = createRef<HTMLDivElement>();
+  const dropdownRef = createRef<HTMLDivElement>();
+  const [isMenuVisible, setIsMenuVisible] = useState(false);
+  const [actionsState, setActionsState] = useState(
+    actions.map((action) => ({ isVisible: true, action: excludeIconProps(action), ref: createRef<HTMLDivElement>() })),
+  );
+
+  const changeActionsVisibility = useCallback(() => {
+    const parentWidth = parentRef.current?.offsetWidth;
+    const dropdownWidth = dropdownRef.current?.offsetWidth;
+
+    if (!parentWidth || !dropdownWidth) {
+      return;
+    }
+
+    let remainingWidth = parentWidth;
+
+    const nextState = actionsState.map((stateObj) => {
+      const actionWidth = stateObj.ref.current?.offsetWidth;
+
+      if (!actionWidth) {
+        return stateObj;
+      }
+
+      if (stateObj.action.actionType === 'destructive') {
+        return { ...stateObj, isVisible: false };
+      }
+
+      if (remainingWidth - actionWidth > dropdownWidth) {
+        remainingWidth = remainingWidth - actionWidth;
+
+        return { ...stateObj, isVisible: true };
+      }
+
+      return { ...stateObj, isVisible: false };
+    });
+
+    const visibleActions = actionsState.filter(({ isVisible }) => isVisible);
+    const nextVisibleActions = nextState.filter(({ isVisible }) => isVisible);
+
+    if (visibleActions.length !== nextVisibleActions.length) {
+      setActionsState(nextState);
+    }
+  }, [actionsState, dropdownRef, parentRef]);
+
+  const renderedDropdown = useMemo(
+    () => (
+      <StyledFlexItem data-testid="button-group-dropdown" isVisible={isMenuVisible} ref={dropdownRef} role="listitem">
```

Comment:
These are the only reliable way we could think of for targeting them and setting the offsetWidths so that we could test the collapsing functionality (jsdom does not actually render things, so all elements have an offsetWidth of 0 in the testing environment)

Reply:
Yeah, that was the same issue we ran into with the PillsTabs.

🍹 I'm :gucci: with this, but could we align it with a similar name to the `PillsTabs` one? Definitely a nitpick:
```suggestion
      <StyledFlexItem data-testid="buttongroup-dropdown-toggle" isVisible={isMenuVisible} ref={dropdownRef} role="listitem">
```

Reply:
Thank you for your comment! Sure, I agree with you!

Done ✅